### DISCUSSION TITLE
xavier gpu not functioning

### DISCUSSION THREAD

**Discussion by lfdmn** [OE4T/meta-tegra#535] [Q&A] [answered]
Hi,

I just upgraded from zeus to dunfell-l4t-r32.4.3.

I have issues with the GPU.

Both jetson_clocks and nvpmodel fail to access the GPU

```
jetson-xavier  # nvpmodel -m 0
NVPM ERROR: Error opening /sys/devices/17000000.gv11b/devfreq/17000000.gv11b/available_frequencies: 2
NVPM ERROR: failed to read PARAM GPU: ARG FREQ_TABLE: PATH /sys/devices/17000000.gv11b/devfreq/17000000.gv11b/available_frequencies
NVPM ERROR: failed to set power mode!
NVPM ERROR: optMask is 1, no request for power mode
```

```
jetson-xavier cortexdaemon # jetson_clocks --show
SOC family:tegra194  Machine:Jetson-AGX
Online CPUs: 0-7
CPU Cluster Switching: Disabled
cpu0: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu1: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu2: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu3: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu4: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu5: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu6: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
cpu7: Online=1 Governor=performance MinFreq=1190400 MaxFreq=2265600 CurrentFreq=2265600 IdleStates: C1=0 c6=0 
/usr/bin/jetson_clocks: line 248: /sys/class/devfreq/*/device/of_node/name: No such file or directory
Error! Unknown GPU!
```

I built the kernel using tegra_defconfig and added some fragments to disable/remove bluetooth, wifi, etc and applied the preempt rt patches. 

dmesg shows some errors for nvgpu: 17000000.gv11b when using nvpmodel
[dmesg.txt](https://github.com/OE4T/meta-tegra/files/5821581/dmesg.txt)

I get some segmentation fault for programs trying to use CUDA.

Does this ring a bell to someone?

Thanks

**Comment by madisongh** [comment]
I've seen something similar.  IIRC, the GPU driver doesn't populate the sysfs entries for frequency control immediately on loading; it has to be brought out of power management suspended state first... or something like that, at least.  In some versions of the BSP, I've also had to forcibly start up something that uses the GPU before the sysfs entries would get populated, although I think that 32.4.3 has been better about that.   Take a look at the `/usr/sbin/nvphsd_setup.sh` script - that might provide some hints.

**Reply by lfdmn** [reply]
Thanks @madisongh that put me to the right track!

```nvphd``` was missing from my images because I was installing ```packagegroup-core-boot``` (```IMAGE_INSTALL = "packagegroup-core-boot ...```) 

but
```
MACHINE_EXTRA_RDEPENDS = "tegra-nvpmodel tegra-nvphs tegra-nvs-service tegra-nvstartup tegra-configs-udev"
MACHINE_EXTRA_RRECOMMENDS = "kernel-module-nvgpu kernel-module-tegra-udrm" 
```
only get installed in the image with ```packagegroup-base``` (```IMAGE_INSTALL = "packagegroup-base ...```).

The implementation was different with Zeus so I didn't have that issue.

After removing my fragments I can control the power model etc. So I go back to re-configuring the kernel and see what screws things up =)

Thanks!

**Comment by lfdmn** [comment, marked as answer]
**Do not** disable scsi drivers in your kernel config

```
# CONFIG_BLK_DEV_SD is not set
# CONFIG_SCSI_LOWLEVEL is not set
```
This was the second cause for the failure.
